### PR TITLE
GH-43946: [C++][Parquet] Guard against use of cleared decryptor/encryptor

### PR DESCRIPTION
**Pull request by pitrou**
This is to get a clearer error rather than an obscure crash, see https://github.com/apache/arrow/issues/43057 for an example.

* GitHub Issue: #43946

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Comment by pitrou** [comment]
@adamreeve Does this look good to you?

**Comment by pitrou** [comment]
We have a nice example of the guard being triggered here:
https://github.com/ursacomputing/crossbow/actions/runs/10699204038/job/29660230356#step:7:3575

This was a sporadic crash before:
https://github.com/apache/arrow/issues/43057

**Review by adamreeve — APPROVED** [review]

**Comment by adamreeve** [comment]
This looks good to me, impressive work tracking down the cause of the segfault!

**Comment by wgtmac** [comment]
> We have a nice example of the guard being triggered here: https://github.com/ursacomputing/crossbow/actions/runs/10699204038/job/29660230356#step:7:3575
> 
> This was a sporadic crash before: #43057

Does it mean that `DatasetEncryptionTest.WriteReadDatasetWithEncryption` test case does not run in the regular CI? Or is it nondeterministic when the decryptor is cleared?

File: cpp/src/parquet/encryption/encryption_internal.cc [hunk 1]
```diff
@@ -413,6 +423,12 @@ class AesDecryptor::AesDecryptorImpl {
   }
 
  private:
+  void CheckValid() {
+    if (ctx_ == nullptr) {
```
Comment (wgtmac) [inline cpp/src/parquet/encryption/encryption_internal.cc:427]:
```suggestion
    if (ARROW_PREDICT_FALSE(ctx_ == nullptr)) {
```

File: cpp/src/parquet/encryption/encryption_internal.cc [hunk 2]
```diff
@@ -89,6 +89,12 @@ class AesEncryptor::AesEncryptorImpl {
...
```
Comment (wgtmac) [inline cpp/src/parquet/encryption/encryption_internal.cc:93]:
```suggestion
    if (ARROW_PREDICT_FALSE(ctx_ == nullptr)) {
```

**Review by wgtmac — APPROVED** [review]

File: cpp/src/parquet/encryption/encryption_internal.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/parquet/encryption/encryption_internal.cc:427]: I was under the impression that compilers automatically treated branches that throw an exception to be unlikely, but I can't find a reference. @felipecrv Do you know about that?

**Comment by pitrou** [comment]
> Does it mean that `DatasetEncryptionTest.WriteReadDatasetWithEncryption` test case does not run in the regular CI? Or is it nondeterministic when the decryptor is cleared?

It seems to be non-deterministic in this particular test, as the dataset reading is multi-threaded (you can see this in the stack traces above).

File: cpp/src/parquet/encryption/encryption_internal.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/parquet/encryption/encryption_internal.cc:427]: Ok, judging from https://www.youtube.com/watch?v=T84swS6DCRo (at 29:00), the gcc compiler already uses such a heuristic.

Comment (mapleFU) [reply cpp/src/parquet/encryption/encryption_internal.cc:427]: CheckValid should be `const`?

Comment (pitrou) [reply cpp/src/parquet/encryption/encryption_internal.cc:427]: Good point @mapleFU

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp

**Pull request merged**